op_ref false positive with deref coercions when comparing

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->

Perhaps related to #2042, but it seemed different enough to open a new issue.

With the following code: https://play.rust-lang.org/?gist=35d367db95cd56394d0bb9a05247b94c&version=nightly

```rust
fn main() {
    let a: &str = "abc";
    let b: String = "abc".to_owned();
    println!("{}", a > &b);
}
```

Clippy complains that the reference should be removed, but removing the reference causes the code to stop compiling.

I just ran into this with a newtype I defined:

```
warning: taken reference of right operand
   --> src/lib.rs:208:37
    |
208 |         || f64::from(stats.words) < f64::from(stats.urls) * &*config.min_words_per_url
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^--------------------------
    |                                                             |
    |                                                             help: use the right value directly: `*config.min_words_per_url`
    |
    = note: `#[warn(clippy::op_ref)]` implied by `#[warn(clippy::all)]`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#op_ref
```

```
error[E0277]: cannot multiply `f64` by `config::newtypes::PositiveF64`
   --> src/lib.rs:208:59
    |
208 |         || f64::from(stats.words) < f64::from(stats.urls) * config.min_words_per_url
    |                                                           ^ no implementation for `f64 * config::newtypes::PositiveF64`
    |
    = help: the trait `std::ops::Mul<config::newtypes::PositiveF64>` is not implemented for `f64`
    = help: the following other types implement trait `std::ops::Mul<Rhs>`:
              <&'a f32 as std::ops::Mul<f32>>
              <&'a f64 as std::ops::Mul<f64>>
              <&'a i128 as std::ops::Mul<i128>>
              <&'a i16 as std::ops::Mul<i16>>
              <&'a i32 as std::ops::Mul<i32>>
              <&'a i64 as std::ops::Mul<i64>>
              <&'a i8 as std::ops::Mul<i8>>
              <&'a isize as std::ops::Mul<isize>>
            and 49 others
```

Granted, I'll implement `Mul` after posting this, but I'm sure there are situations where that's not an option.

--- a/clippy_lints/src/operators/op_ref.rs
+++ b/clippy_lints/src/operators/op_ref.rs
@@ -86,7 +86,7 @@ pub(crate) fn check<'tcx>(
                                 left.span,
                                 "use the left value directly",
                                 lsnip,
-                                Applicability::MaybeIncorrect, // FIXME #2597
+                                Applicability::MachineApplicable,
                             );
                         },
                     );
@@ -105,7 +105,7 @@ pub(crate) fn check<'tcx>(
                                 right.span,
                                 "use the right value directly",
                                 rsnip,
-                                Applicability::MaybeIncorrect, // FIXME #2597
+                                Applicability::MachineApplicable,
                             );
                         },
                     );
@@ -137,7 +137,7 @@ pub(crate) fn check<'tcx>(
                                 left.span,
                                 "use the left value directly",
                                 lsnip,
-                                Applicability::MaybeIncorrect, // FIXME #2597
+                                Applicability::MachineApplicable,
                             );
                         },
                     );
@@ -164,7 +164,7 @@ pub(crate) fn check<'tcx>(
                             right.span,
                             "use the right value directly",
                             rsnip,
-                            Applicability::MaybeIncorrect, // FIXME #2597
+                            Applicability::MachineApplicable,
                         );
                     });
                 }

--- a/tests/ui/op_ref.fixed
+++ b/tests/ui/op_ref.fixed
@@ -98,3 +98,15 @@ impl Mul<A> for A {
         self * &rhs
     }
 }
+
+mod issue_2597 {
+    fn ex1() {
+        let a: &str = "abc";
+        let b: String = "abc".to_owned();
+        println!("{}", a > &b);
+    }
+
+    pub fn ex2<T: Ord + PartialOrd>(array: &[T], val: &T, idx: usize) -> bool {
+        &array[idx] < val
+    }
+}

--- a/tests/ui/op_ref.rs
+++ b/tests/ui/op_ref.rs
@@ -98,3 +98,15 @@ impl Mul<A> for A {
         self * &rhs
     }
 }
+
+mod issue_2597 {
+    fn ex1() {
+        let a: &str = "abc";
+        let b: String = "abc".to_owned();
+        println!("{}", a > &b);
+    }
+
+    pub fn ex2<T: Ord + PartialOrd>(array: &[T], val: &T, idx: usize) -> bool {
+        &array[idx] < val
+    }
+}

